### PR TITLE
Subscription for mainnet

### LINKED ISSUE #20: Subscription for mainnet
Once #19 is ready, when creating a mainnet wallet it should prompt for a subscription.

When cancelling a subscription, the wallet should remain available, but the transaction Sign button should prompt for a new subscription. The user can always export their mnemonic and use a different (hardware) wallet to sign, so they are never locked out of their coins.

### PR DESCRIPTION
**Pull request by w-i-n-s**
resolve #20

### CODE REVIEW

**Comment by Sjors** [comment]
Some UI suggestions (I know this PR is still draft):

Schermafbeelding 2021-05-03 om 10 50 42


Let's disable the Mainnet tabs under "Import" and Announce. Then, under misc, add "Mainnet Subscription". This should open a new view (`SubscriptionView.swift`?) where we can show the status and explain the subscription. I believe this also needs to have a "restore existing purchase" button according to the app store rules?

**Comment by w-i-n-s** [comment]
@Sjors 
> Let's disable the Mainnet tabs under "Import" and "Announce". Then, under misc, add "Mainnet Subscription". This should open a new view (`SubscriptionView.swift`?) where we can show the status and explain the subscription. I believe this also needs to have a "restore existing purchase" button according to the app store rules?
> 
> In the list of wallets, don't disable mainnet wallets. In the sign screen, the sign button should be disabled and there should be a link to `SubscriptionView`.

UI and navigation was fixed in d35bb2cf21ab61442fab04014769a539e11362f4

File: NthKey.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -975,7 +975,7 @@
 				DEVELOPMENT_TEAM = 227MFQPYQ4;
 				ENABLE_PREVIEWS = YES;
 				INFOPLIST_FILE = "$(SRCROOT)/NthKey/SupportingFiles/Info.plist";
-				IPHONEOS_DEPLOYMENT_TARGET = 13.2;
+				IPHONEOS_DEPLOYMENT_TARGET = 14.1;
```
Comment (Sjors) [inline NthKey.xcodeproj/project.pbxproj:982]: This breaks the Travis test, maybe another place needs the version bump too? https://travis-ci.com/github/Sjors/nthkey-ios/builds/226255791#L1477

File: NthKey/App/SceneDelegate.swift [hunk 2]
```diff
@@ -14,7 +14,7 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
 
     private let store = PersistentStore(inMemory: false)
     private lazy var dataManager = DataManager(store: store)
-    private lazy var subsManager = SubscriptionManager()
+    private lazy var subsManager = SubscriptionManager(identifiers: ["com.nthkey.monthly", "com.nthkey.annual"])
```
Comment (Sjors) [inline NthKey/App/SceneDelegate.swift:17]: I'd like to start with just monthly. We'll add annual later.

File: NthKey.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Sjors) [reply NthKey.xcodeproj/project.pbxproj:982]: Also, if you're bumping the iOs version, let's also update the recommended project settings and fix the deprecation warnings.

**Comment by Sjors** [comment]
It's difficult to dismiss this modal (at least in the simulator). Maybe make it smaller?
Schermafbeelding 2021-05-19 om 19 58 54

**Comment by Sjors** [comment]
I get a crash when clicking Subscribe Now:
Schermafbeelding 2021-05-19 om 20 00 19

File: NthKey/Model/SubscriptionManager.swift [hunk 3]
```diff
@@ -7,13 +7,240 @@
...
+        request.start()
+    }
+
+    private func checkPurchaseStatus() {
+        guard let date = UserDefaults.subscriptionDate else { return }
+        hasSubscription = Date() < date
+    }
+
+    /// Bad practice apple.com/documentation/storekit/in-app_purchase/validating_receipts_with_the_app_store
+    /// Do not call the App Store server verifyReceipt endpoint from your app...
```
Comment (Sjors) [inline NthKey/Model/SubscriptionManager.swift:79]: Can you make an issue for this? We can fix it after the next TestFlight release, but maybe before the App Store release.

**Comment by w-i-n-s** [comment]
> It's difficult to dismiss this modal (at least in the simulator). Maybe make it smaller?

Added `Close` button 37f91216a947680e449d85f28ada1d14e15b403f

File: NthKey/Model/SubscriptionManager.swift (hunk 3 above)
Comment (w-i-n-s) [reply NthKey/Model/SubscriptionManager.swift:79]: added https://github.com/Sjors/nthkey-ios/issues/46

**Comment by w-i-n-s** [comment]
> I get a crash when clicking Subscribe Now:

It happens because products didn't receive before you tap on the button. I'll implement FSM to avoid such issues

File: NthKey/App/SceneDelegate.swift (hunk 2 above)
Comment (w-i-n-s) [reply NthKey/App/SceneDelegate.swift]: fixed in 9b52487b5ce2be3649791267d10e26b385a4aa0a

File: NthKey.xcodeproj/project.pbxproj (hunk 1 above)
Comment (w-i-n-s) [reply NthKey.xcodeproj/project.pbxproj:994]: fixed in 5b33af1b0cb99302d813be9ee138140f1ad818d8

File: NthKey/View/Subscription/SubscriptionView.swift [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    .foregroundColor(.gray)
+                    .padding()
+            })
+        }
+
+        return NavigationView {
+            Group {
+                switch model.state {
+                case .initial:
+                    Text("Request AppStore")
```
Comment (Sjors) [inline NthKey/View/Subscription/SubscriptionView.swift:74]: "Connecting to App Store"

**Review by Sjors — COMMENTED** [review]
Can you squash 5b33af1 into b9c3ff9, and squash 9b52487 into 96d17d91256d92ea4811a086706eb344b10399b4?

Both the simulator and device are stuck for me here as soon as I click on Mainnet (no error in console):
Schermafbeelding 2021-05-21 om 11 32 04

On the send screen, if I open an existing mainnet wallet, I see a console message:
> Next products are not on the store anymore:["com.nthkey.monthly"]
> request did finish

Let's also drop the `validateReceipt` code in this PR (see #46); that leaves less review work.

File: NthKey/Model/Configuration.storekit [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          "internalID" : "F3FCF59A",
+          "introductoryOffer" : null,
+          "localizations" : [
+            {
+              "description" : "",
+              "displayName" : "",
+              "locale" : "en_US"
+            }
+          ],
+          "productID" : "com.nthkey.annual",
```
Comment (Sjors) [inline NthKey/Model/Configuration.storekit:66]: Looks like you forgot remove this in 9b52487.

File: NthKey/Model/Configuration.storekit [hunk 6]
```diff
@@ -0,0 +1,79 @@
+{
+  "products" : [
+
+  ],
+  "settings" : {
+    "_timeRate" : 0
+  },
+  "subscriptionGroups" : [
+    {
+      "id" : "5BC28C3C",
```
Comment (Sjors) [inline NthKey/Model/Configuration.storekit:10]: Where do these identifiers come from?

File: NthKey/Model/Configuration.storekit [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+      ],
+      "name" : "subscription_group",
+      "subscriptions" : [
+        {
+          "adHocOffers" : [
+            {
+              "internalID" : "F0D0AE94",
+              "offerID" : "",
+              "paymentMode" : "free",
```
Comment (Sjors) [inline NthKey/Model/Configuration.storekit:21]: Let's drop the free trial; the user can already use testnet and signet for free.
(even if this is just for testing, it reduces the amount of code which is always nice)

File: NthKey/Model/SubscriptionManager.swift [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                    checkExpirationDateFromPayment(transaction.payment)
+                }
+            case .failed:
+                if let error = transaction.error as? SKError {
+                    purchasePublisher.send(("Payment Error \(error.code) ",false))
+                    print("Payment Failed \(error.code)")
+                }
+            case .deferred:
+                print("Ask Mom ...")
+                purchasePublisher.send(("Payment Diferred ",false))
```
Comment (Sjors) [inline NthKey/Model/SubscriptionManager.swift:187]: "Payment Deferred " (don't forget to drop these `print` debug lines everywhere)

File: NthKey/Model/SubscriptionManager.swift [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+    func request(_ request: SKRequest, didFailWithError error: Error) {
+        purchasePublisher.send(("Purchase request failed ",true))
+        print("didFailWithError ",error)
+    }
+}
+
+#if DEBUG
+extension SubscriptionManager {
+    static var mock: SubscriptionManager = SubscriptionManager(identifiers: ["com.nthkey.monthly", "com.nthkey.annual"])
```
Comment (Sjors) [inline NthKey/Model/SubscriptionManager.swift:237]: Forgot to drop annual here?

File: NthKey/Reusables/Extensions/SKProduct.swift [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+extension SKProductSubscriptionPeriod {
+    func localizedPeriod() -> String? {
+        return PeriodFormatter.format(unit: unit.toCalendarUnit(), numberOfUnits: numberOfUnits)
+    }
+}
+
+extension SKProductDiscount {
+    func localizedDiscount() -> String? {
+        switch paymentMode {
+        case PaymentMode.freeTrial:
```
Comment (Sjors) [inline NthKey/Reusables/Extensions/SKProduct.swift:81]: Free trial handling can be dropped.

**Comment by Sjors** [comment]
On the sign screen, if the user selects the mainnet wallet, and they are not (or no longer) subscribed, let's disable the "scan psbt" and  "load psbt" buttons and display a link: "A subscription is required to sign with this wallet". Clicking that opens the modal.

**Comment by Sjors** [comment]
Some minor issues:
Schermafbeelding 2021-05-24 om 17 53 22

The modal should be closed when a purchase is successful.

File: NthKey/Model/Configuration.storekit (hunk 5 above)
Comment (w-i-n-s) [reply NthKey/Model/Configuration.storekit]: This is test configuration which didn't affect to real iAP

File: NthKey/Model/Configuration.storekit (hunk 6 above)
Comment (w-i-n-s) [reply NthKey/Model/Configuration.storekit:14]: It set/appears automatically for Identifiable protocol support

File: NthKey/Model/SubscriptionManager.swift (hunk 9 above)
Comment (w-i-n-s) [reply NthKey/Model/SubscriptionManager.swift]: No, we can ask AppStore for any iAP products (it used for debug and preview). If some of product isn't in the store now then we'll receive info about it in debug https://github.com/Sjors/nthkey-ios/blob/8feb68d44d7544e8506ce74dfc68756ebcb703e2/NthKey/Model/SubscriptionManager.swift#L103

**Comment by w-i-n-s** [comment]
> Can you squash [5b33af1](https://github.com/Sjors/nthkey-ios/commit/5b33af1b0cb99302d813be9ee138140f1ad818d8) into [b9c3ff9](https://github.com/Sjors/nthkey-ios/commit/b9c3ff979e5c536e28bc7651a75e5ac24b412c4d), and squash [9b52487](https://github.com/Sjors/nthkey-ios/commit/9b52487b5ce2be3649791267d10e26b385a4aa0a) into [96d17d9](https://github.com/Sjors/nthkey-ios/commit/96d17d91256d92ea4811a086706eb344b10399b4)?
> 


> Both the simulator and device are stuck for me here as soon as I click on Mainnet (no error in console):
> Schermafbeelding 2021-05-21 om 11 32 04
> 
> On the device (which has an existing mainnet wallet) I see a console message:
> 
> > Next products are not on the store anymore:["com.nthkey.monthly"]
> > request did finish
> 
> Let's also drop the `validateReceipt` code in this PR (see #46); that leaves less review work.

**Comment by Sjors** [comment]
I was able to make a purchase using the sandbox user. However the "Use real Bitcoin (mainnet)" dialog doesn't close when the purchase is complete.

I also suggest making the "Subscribe Now" button grey while the purchase is in progress.

Both these issues can wait for a future PR if they're not simple.

**Pull request merged**